### PR TITLE
Blossom Implement

### PR DESCRIPTION
**Pull request by zhaodice**
## Description

Please carefully read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md) before making any pull requests.
And, **Do not make a pull request to merge into stable unless it is a hotfix. Use the development branch instead.**
## Type of changes

- [ ] Bug fix
- [x] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.
image

### CODE REVIEW

**Comment by Birdulon** [comment]
Notes from testing:
Map icons don't seem to disappear until teleporting away
image

Resin doesn't seem to be consumed when claiming rewards, neither original resin nor condensed resin
image

Condensed resin doesn't give double reward
image

That's a lot of ley lines
image

**Comment by zhaodice** [comment]
> Notes from testing: Map icons don't seem to disappear until teleporting away image
> 
> Resin doesn't seem to be consumed when claiming rewards, neither original resin nor condensed resin image
> 
> Condensed resin doesn't give double reward image
> 
> That's a lot of ley lines image

icon will delay to load because of spawn logic, all of blossom will be loaded because I doesn't implement refresh logic which should be implemented with big world collection together

**Comment by zhaodice** [comment]
double reward implement is coming soon

File: src/main/java/emu/grasscutter/data/GameDepot.java
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/data/GameDepot.java]: In Java, it's always best practice to use the base interface of a 'generic' object. (there's some fancy term for it but to sum it up) In this instance, the property type declaration would be `Map<Integer, List<BlossomReward>>`.

Comment (zhaodice) [reply src/main/java/emu/grasscutter/data/GameDepot.java]: resolved

**Comment by zhaodice** [comment]
@KingRainbow44 All done, I tested it and it is no problem

**Comment by zhaodice** [comment]
condensed resin supported:
image

**Review by KingRainbow44 — APPROVED** [review]

**Comment by eternalcomet** [comment]
you need reviews from @Birdulon @Melledy ,but why don't they approve?you can try to cancel "request a review from @Birdulon @Melledy "

**Comment by zhaodice** [comment]
@eternalcomet melledy is busy, it doesn't hurry

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomActivity.java
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomActivity.java]: reward?

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java]:
```suggestion
```

Big commented block

File: src/main/java/emu/grasscutter/server/packet/send/PacketBlossomBriefInfoNotify.java
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/send/PacketBlossomBriefInfoNotify.java]:
```suggestion
```

Big commented block

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomType.java
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomType.java]:
```suggestion
        return switch(i) {
            case GOLDEN_GADGET_ID.gadgetId -> GOLDEN_GADGET_ID;
            case BLUE_GADGET_ID.gadgetId -> BLUE_GADGET_ID;
            default -> null;
        };
```

**Review by Birdulon — CHANGES_REQUESTED** [review]

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomType.java
Comment (Birdulon) [reply src/main/java/emu/grasscutter/game/managers/blossom/BlossomType.java]: nevermind, they aren't static values

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:137]: This feels like it's going around in circles but I can't be bothered to confirm if it's safe to just delete this and use the raw `SpawnDataEntry` stream.

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:44]: Not a fan of these iterator patterns, not sure if there's a cleaner way of doing iterate and delete though.

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:197]: Not a fan of these iterator patterns, not sure if there's a cleaner way of doing find and delete though.

**Review by Birdulon — APPROVED** [review]
Tested locally, mostly works. There's a few things that need some more attention but they can probably wait until after merging this set of changes.

Leaving a few comments up for @Melledy to look over.

To all other write-access: **Please don't merge this without squashing!**

**Comment by eternalcomet** [comment]
maybe we can use these files to control the logic
```
"\resources\ExcelBinOutput\BlossomReviseExcelConfigData.json"
"\resources\ExcelBinOutput\BlossomSectionOrderExcelConfigData.json"
"\resources\ExcelBinOutput\BlossomTalkExcelConfigData.json"
"\resources\ExcelBinOutput\BlossomChestExcelConfigData.json"
"\resources\ExcelBinOutput\BlossomGroupsExcelConfigData.json"
"\resources\ExcelBinOutput\BlossomOpenExcelConfigData.json"
"\resources\ExcelBinOutput\BlossomRefreshExcelConfigData.json"(have been used by this pr)
```

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomActivity.java
Comment (eternalcomet) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomActivity.java:125]: we may read resin cost and chestGadgetId from `"\resources\ExcelBinOutput\BlossomChestExcelConfigData.json"`instead of hard coding

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (eternalcomet) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java]: we may rename "leyline" to "blossom"(official name)

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (eternalcomet) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:79]: we should check open state for the player(level>2),and the level should be read from file `BlossomOpenExcelConfigData.json`

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (eternalcomet) [inline src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java]: rename leyline to blossom

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (Birdulon) [reply src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:79]: Actually the unlock conditions are under `refreshCondVec` in `Resources/ExcelBinOutput/BlossomRefreshExcelConfigData.json` - it's AR8 for the Mondstadt EXP leyline, AR12 for Mondstadt mora leyline, AR18 for both Liyue leylines, AR30 for both Inazuma leylines.

Copying the official spawns will be a bit more work after this PR.

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (zhaodice) [reply src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:137]: > This feels like it's going around in circles but I can't be bothered to confirm if it's safe to just delete this and use the raw `SpawnDataEntry` stream.

emmm, see follows?

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (zhaodice) [reply src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:197]: > Not a fan of these iterator patterns, not sure if there's a cleaner way of doing find and delete though.

The iterator could help us safely delete item, because when I use for with index to delete item, the index will change when I delete a item,program will crash immediately

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomActivity.java
Comment (zhaodice) [reply src/main/java/emu/grasscutter/game/managers/blossom/BlossomActivity.java:125]: > we may read resin cost and chestGadgetId from `"\resources\ExcelBinOutput\BlossomChestExcelConfigData.json"`instead of hard coding

oh, I had overlooked that...

**Comment by zhaodice** [comment]
> maybe we can use these files to control the logic
> 
> ```
> "\resources\ExcelBinOutput\BlossomReviseExcelConfigData.json"
> "\resources\ExcelBinOutput\BlossomSectionOrderExcelConfigData.json"
> "\resources\ExcelBinOutput\BlossomTalkExcelConfigData.json"
> "\resources\ExcelBinOutput\BlossomChestExcelConfigData.json"
> "\resources\ExcelBinOutput\BlossomGroupsExcelConfigData.json"
> "\resources\ExcelBinOutput\BlossomOpenExcelConfigData.json"
> "\resources\ExcelBinOutput\BlossomRefreshExcelConfigData.json"(have been used by this pr)
> ```

that looks like a huge Engineering,hmmm.. I think this task could move to enhancement plan

File: src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java
Comment (zhaodice) [reply src/main/java/emu/grasscutter/game/managers/blossom/BlossomManager.java:44]: > Not a fan of these iterator patterns, not sure if there's a cleaner way of doing iterate and delete though.

The code is designed to locate blossoms in the WHOLE map,so I have to use GameDepot.getSpawnLists() , if you have a good idea ,feel free to improve that

**Comment by eternalcomet** [comment]
so maybe it can be merged?

**Comment by Birdulon** [comment]
> so maybe it can be merged?

Meledy has expressed interest in personally reviewing it. In the meantime, loading enemy spawn groups from `BlossomGroupsExcelConfigData.json` and blossom spawn patterns from `BlossomSectionOrderExcelConfig.json` to bring things in line with official server behaviour would be a good goal.

**Review by ghost — APPROVED** [review]

**Pull request merged**
